File: python/twelve-days/twelve_days.py
gifts = [
    ("first", "a Partridge in a Pear Tree."),
    ("second", "two Turtle Doves"),
    ("third", "three French Hens"),
    ("fourth", "four Calling Birds"),
    ("fifth", "five Gold Rings"),
    ("sixth", "six Geese-a-Laying"),
    ("seventh", "seven Swans-a-Swimming"),
    ("eighth", "eight Maids-a-Milking"),
    ("ninth", "nine Ladies Dancing"),
    ("tenth", "ten Lords-a-Leaping"),
    ("eleventh", "eleven Pipers Piping"),
    ("twelfth", "twelve Drummers Drumming"),
]
def recite(start_verse, end_verse):
    out = []
    for i in range(start_verse - 1, end_verse):
        gift = gifts[i]
        line = 'On the ' + gift[0] + ' day of Christmas my true love gave to me: ' + gift[1]
        if i > 0:
            line += ', '
            extra = [g[1] for g in gifts[i - 1::-1]]
            extra[-1] = 'and ' + extra[-1]
            line += ', '.join(extra)
        out.append(line)
    return out

File: ruby/twelve-days/twelve_days.rb
class TwelveDays
  PHRASE = "On the %s day of Christmas my true love gave to me: "
  VERSES = ['a Partridge in a Pear Tree.', 'two Turtle Doves', 'three French Hens', 'four Calling Birds', 'five Gold Rings', 'six Geese-a-Laying', 'seven Swans-a-Swimming', 'eight Maids-a-Milking', 'nine Ladies Dancing', 'ten Lords-a-Leaping', 'eleven Pipers Piping', 'twelve Drummers Drumming']
  COUNTERS = %w[first second third fourth fifth sixth seventh eighth ninth tenth eleventh twelfth]
  def self.song
    song = PHRASE % [COUNTERS[0]] + VERSES[0] + "\n\n"
    former = ['and ' + VERSES[0]]
    (1..11).each do |idx|
      former.unshift(VERSES[idx])
      song += PHRASE % [COUNTERS[idx]] + former.join(', ')
      song += "\n\n" unless idx == 11
    end
    song + "\n"
  end
end
